filewatcher/filewatcher.go

```diff
@@ -379,3 +396,19 @@ func (fw *MockFileWatcher) Get() interface{} {
 func (fw *MockFileWatcher) Stop() {}
 
 var _ FileWatcher = (*MockFileWatcher)(nil)
+
+// DummyReadCloser is a mock struct used to hold the path for directory watchers
+type DummyReadCloser struct {
```

Comment:
don't export this

Reply:
I export this so I can use it to type assert in secrets/store.go

Reply:
I think what we need here is to:
1. Define a type, `DirParser`, for use cases with a directory
2. Provide a wrapper to wrap `DirParser` into `Parser` so they can be used in the current API

and secrets should use their own `DirParser`, which should be roughly the same as the callback you are currently using in `Walk`.

Reply:
actually the `DirParser` api should be something like this:

```go
type DirParer func(path string) (data interface{}, err error)
```

then we provide another api to convert that into a `Parser`:

```go
func DirParserToParser(dp DirParser) Parser {
  return func(f io.Reader) (interface{}, error) {
    return dp(f.(dummyReader).Path)
  }
}
```

and we no longer need to export `DummyReaderCloser`. `Walk` should be implemented inside `DirParser` instead.

Reply:
ah nice, thank you

Reply:
I'm still interested in seeing this approach considered, as I still consider mis-implementing an interface to be a loaded footgun we don't want to saddle ourselves with